### PR TITLE
Cover: Customizing Alignment of inner content

### LINKED ISSUE #20836: Cover: internal alignment / position
The **Cover** block is a pretty nice tool. However, its internal positioning is quite inflexible. You can align individual blocks to left / right / center, but not a group, and not vertically.

It'd be neat to explore some more cardinal positions:

image

Combining them in a single tool should help with clarity:

image

Since there is no internal width or padding (at least until the work on dimensions in "Group" is finalized), you should still be able to flush individual blocks to the left or right even with a default set:

image

### PR DESCRIPTION
**Pull request by ItsJonQ**
## Description

Screen Capture on 2020-03-25 at 17-26-51

This update brings the ability to adjust the vertical and horizontal alignment of content with the Cover block. The control is provided by a new (experimental) `AlignmentMatrixControl` component from the `@wordpress/components` package.

Attempts to resolve: https://github.com/WordPress/gutenberg/issues/20836

### Control

Screen Capture on 2020-03-23 at 16-52-09

The new `AlignmentMatrixControl` provides a "9 point" alignment interface. This UI can be controlled by clicking on any of the points. Alternatively, pressing `UP`, `DOWN`, `LEFT`, or `RIGHT` on the keyboard would change the alignment values.

### Icon

76532112-70d2c080-6476-11ea-8dd0-1fedee597c2f

In addition to the control, the `AlignmentMatrixControl` provides a companion `icon`, which can be loaded into UI like the Toolbar. It accepts an `alignment` prop, which updates the UI to visually reflect the alignment.

The icon implementation is designed to work like an icon from `@wordpress/icons`

##### Example

```jsx
import { __experimentalAlignmentMatrixControl as AlignmentMatrixControl } from "@wordpress/components";
import { Icon } from "@wordpress/icons";

const Example = () => {
  return <Icon icon={AlignmentMatrixControl.icon} alignment="bottom left" />;
};

```

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/components/src/alignment-control/utils.js
Comment (ZebulanStanphill) [inline packages/components/src/alignment-control/utils.js]:
```suggestion
		ALIGNMENT_MATRIX[ x ] !== undefined &&
		ALIGNMENT_MATRIX[ x ][ y ] !== undefined;
```

File: packages/components/src/alignment-control/README.md
Comment (mtias) [inline packages/components/src/alignment-control/README.md]: Should we call this `AlignmentMatrix` to better separate from `BlockAlignment`, etc?

Comment (ItsJonQ) [reply packages/components/src/alignment-control/README.md]: We could! Although I think having `Control` in the component name would be helpful, as it has the `value/onChange` interface.

`AlignmentMatrixControl`? or `AlignmentGridControl`?

**Comment by ItsJonQ** [comment]
Screen Shot 2020-03-24 at 12 32 11 PM

Still very WIP! I have this so far.

In doing so... I'm realizing that it's quite tricky to adjust the size of the Dropdown Menu within this context 🤔 .

**Comment by ItsJonQ** [comment]
Oh boy!!

Screen Capture on 2020-03-25 at 17-26-51

**Comment by ItsJonQ** [comment]
An observation!

In order for non-centered alignment to look okay, I had to add extra padding around the content. That way, the text doesn't touch the edges.

**Comment by pablohoneyhoney** [comment]
Fair! 

It’d be good to test it with an image if reduced within the cover area.

**Comment by mtias** [comment]
@ItsJonQ some default padding makes sense, indeed, but it should be just the natural padding of the "cover" block — [which as part of this work](https://github.com/WordPress/gutenberg/pull/16730) we should make configurable as well.

This is looking great!

**Comment by ItsJonQ** [comment]
> but it should be just the natural padding of the "cover" block

Ah gotcha! Thanks for that feedback @mtias ! I'll make those adjustments now :)

File: packages/block-library/src/cover/style.scss
Comment (ZebulanStanphill) [inline packages/block-library/src/cover/style.scss]: This style belongs with the other general Cover styles at the top of the file, doesn't it?

File: packages/block-library/src/cover/editor.scss
Comment (ZebulanStanphill) [inline packages/block-library/src/cover/editor.scss]: Again, these styles belong at the top of the file. Also, why are we setting the padding in `editor.scss` if we already set it in `style.scss`? And why is the value different here?

**Comment by ItsJonQ** [comment]
Screen Shot 2020-03-27 at 11 12 22 AM

Working on a11y support. After that, will add RTL support!

**Comment by ItsJonQ** [comment]
## a11y + RTL updates

Screen Capture on 2020-03-27 at 11-43-18

The GIF above showcases Mac OS voiceover describing the alignment in when changes are made (keyboard navigation). The control is also rendering in `rtl`, where the labels and values are flipped.

cc'ing @diegohaz I would love your insight and expertise here!

I made an attempt at `aria` labels - I am unsure what `WAI-ARIA` role fits this particular 9x9 alignment UI the best.

@gziolo Also pointed out the [Composite component from Reakit](https://next.reakit.io/docs/composite/) to me. If possible, it would be great to leverage it here 😍 

Thank you so much!

**Comment by diegohaz** [comment]
Great job! It's working really well with keyboard. It can be improved for screen readers though. 

> I am unsure what WAI-ARIA role fits this particular 9x9 alignment UI the best.

It's definitely a [grid](https://www.w3.org/TR/wai-aria-practices/#grid). Listbox is unidirectional. Even though sighted users would see that it's a grid and they could guess they could use arrow up and down to navigate vertically, there's no hint for screen reader users, unless you use `role="grid"` (which is equivalente to `<table>`).

I couldn't reproduce what you show on the GIF using VoiceOver + Chrome and Safari. I can only navigate between the grid items using VO (option+ctrl) + directional keys. And it only navigates horizontally. Using up and down arrow keys (without VO key) only moves to the previous and next elements of the type I selected in the VoiceOver Rotor (by pressing up and right keys fast). For example, if it's set to navigate between links, pressing arrow down would put me in the next link. Maybe it's the MacOS version (10.15.3) or some configuration on VoiceOver?

If that's a `role="grid"`, in combination with `role="row"` and `role="gridcell"`, the screen reader will properly announce the row and column numbers and let you use VO + up or down keys to move vertically.

> @gziolo Also pointed out the Composite component from Reakit to me. If possible, it would be great to leverage it here 😍

That's still experimental, but you can see many implementation examples here, including grid: https://codesandbox.io/s/composite-gcqs2

**Comment by ItsJonQ** [comment]
@diegohaz Wonderful 😍 ! Thank you so much for your feedback! This is tremendously helpful!
I'll move forward with your suggestions

**Comment by ItsJonQ** [comment]
@diegohaz Haii! I started looking into using `Composite`. The API feels great :)

The only thing I'm unsure of is if I'm able to tap into an `onChange` callback of some kind. For example, when I'm using it to display the 9x9 grid, I need to manage state (and pass up) changes.

```jsx
const composite = useCompositeState({
    ...
    onChange: handleOnChange
});
```

Is it possible to do something like this? If not, is there a preferred way of handling this?

Thank you! <3

**Comment by diegohaz** [comment]
@ItsJonQ Depending on what you want to do, you can simply attach an `onFocus` or `onClick` handler to the item:
```jsx
<Composite>
  ...
  <CompositeItem onFocus={select} onClick={select} />
  ...
</Composite>
```

If it's not enough, you can track the `composite.currentId` state:
```jsx
const composite = useCompositeState({ currentId: "unique-id-center-center" });

React.useEffect(() => {
  select(composite.currentId);
}, [composite.currentId]);

<Composite>
  ...
  <CompositeItem id="unique-id-center-center" />
  ...
</Composite>
```

**Comment by ItsJonQ** [comment]
@diegohaz Ooo! Gotcha. Thank you! I'll give that a shot!

**Comment by ItsJonQ** [comment]
Update! I attempted to implement `Composite` into `AlignmentMatrixControl`.
It seemed to work okay, but it's having some issues in the editor.

I've created a separate branch + Draft PR here:
https://github.com/WordPress/gutenberg/pull/21333

**Comment by ItsJonQ** [comment]
Just merged in the a11y work to integrate [Reakit Composite](https://next.reakit.io/docs/composite/) to take care of the interactions and `aria` semantics 🙌 

(Massive props to @diegohaz !)

With that, I think this PR is ready for review

File: packages/components/src/alignment-matrix-control/utils.js
Comment (diegohaz) [inline packages/components/src/alignment-matrix-control/utils.js:38]: It's generating IDs with spaces. Shouldn't it be `replace( ' ', '-' )`?

File: packages/components/src/alignment-matrix-control/utils.js
Comment (diegohaz) [inline packages/components/src/alignment-matrix-control/utils.js:51]: `parseValue` is already replacing `-` by spaces.

File: packages/components/src/alignment-matrix-control/utils.js
Comment (diegohaz) [inline packages/components/src/alignment-matrix-control/utils.js:23]: Maybe it should have a better name to communicate that it's replacing `-` by spaces?

**Review by diegohaz — COMMENTED** [review]
Great job! I tested with Safari + VoiceOver and it's working very well. The VoiceOver cursor ring has a weird position and size due to the visually hidden text, but I guess that's okay.

Apr-03-2020 18-30-14

Also tested with NVDA on Windows, and it also works great.

The only problem I noticed is that IDs are being generated with an space between the two axis. It doesn't seem to affect the usage with screen readers, but I guess it's not valid ID.

File: packages/components/src/alignment-matrix-control/stories/index.js
Comment (ItsJonQ) [reply packages/components/src/alignment-matrix-control/stories/index.js:37]: Nope! Thanks for catching! Will remove

File: packages/components/src/alignment-matrix-control/index.js
Comment (mtias) [inline packages/components/src/alignment-matrix-control/index.js:118]: Should we include some instructions on what it does here? "Align the contents of the block to one of the grid points". Maybe it's unnecessary.

**Comment by mtias** [comment]
@ItsJonQ this is really nice. There are a few design details to polish and we should be good to merge it:

image

- The toolbar group needs to combine the width alignments with the matrix alignment (and the color, in the future).
- The toolbar icon should be slightly smaller (the gaps between the points).
- When the dropdown is not active, can we try gray for the inactive dots and dark gray for the active one?

File: packages/components/src/alignment-matrix-control/index.js
Comment (diegohaz) [inline packages/components/src/alignment-matrix-control/index.js:56]: If items need to show tooltips, I would recommend not to use `aria-activedescendant` here (by removing this `unstable_virtual: true`). It would also make the `handleOnBlur` workaround unnecessary.

**Comment by ItsJonQ** [comment]
@diegohaz + @mtias Thank you both for your feedback! I'm working through the updates now.

P.S. I tried adding `Tooltip` to the matrix cells. When keyboard navigating... this happens:

Screen Shot 2020-04-06 at 11 03 47 AM

I think I'll avoid Tooltip for now 😂

**Comment by mtias** [comment]
Ugh, why does that happen?

**Comment by diegohaz** [comment]
@ItsJonQ I'm pretty sure that's because of `aria-activedescendant`. If you remove `unstable_virtual: true` from `useCompositeState`, that should work.

File: packages/components/src/alignment-matrix-control/index.js
Comment (ItsJonQ) [reply packages/components/src/alignment-matrix-control/index.js:56]: That works @diegohaz !! Thank you 🙏

**Comment by ItsJonQ** [comment]
I was able to apply the fix to enable Tooltips (thanks @diegohaz !)
Now to group the Toolbar items together (which is proving to be tricky)

**Comment by mtias** [comment]
We could look at that separately if it's more convoluted.

**Comment by ItsJonQ** [comment]
> We could look at that separately if it's more convoluted.

I think that would be better! We can a follow up that improves the Toolbar rendering and perhaps (style) attribute registration.

Maybe something like:

```
export const settings = {
	title: __( 'Cover' ),
	...
	supports: {
		align: true,
		html: false,
		__experimentalContentPosition: true,
	},
	...
```

Which would automatically render the control into the Toolbar.
At the moment, the Control is manually added in the Cover's `edit.js`.

**Comment by ItsJonQ** [comment]
@mtias + @pablohoneyhoney 👋 Haii!

No rush! I just wanted to check if you folks were okay merging this PR as it is right now :).
I'm happy to make any adjustments if needed!

Thank you! 🙏

**Comment by mtias** [comment]
@ItsJonQ yes, let's get it in! We can keep refining the experience later.

**Comment by ItsJonQ** [comment]
> yes, let's get it in! We can keep refining the experience later.

@mtias Awesome! Thank you!! If I could get a review and 👍 (from you, @youknowriad , or anyone else), we can merge it 🙌 

(I've rebased it with latest `master`)

File: packages/block-editor/src/components/block-alignment-matrix-toolbar/index.js
Comment (youknowriad) [inline packages/block-editor/src/components/block-alignment-matrix-toolbar/index.js:14]: I know we have this tendency to make a lot of things experimental but I'd like to ask what level of confidence we have over the naming and the props (API) of this component. If it's a high level, maybe we can just make it stable?

File: packages/block-editor/src/components/block-alignment-matrix-toolbar/style.scss
Comment (youknowriad) [inline packages/block-editor/src/components/block-alignment-matrix-toolbar/style.scss:6]: Curious to know why we need custom styling and whether these could be abstracted somehow in Dropdown.

File: packages/block-editor/src/components/index.js
Comment (youknowriad) [inline packages/block-editor/src/components/index.js:11]: Same question about the "experimental" aspect.

File: packages/components/src/alignment-matrix-control/icon.js
Comment (youknowriad) [inline packages/components/src/alignment-matrix-control/icon.js:26]: Feels like there's too much props for a public API component. Do we really need all these configs? I'd have thought we need just "value" and potentially "size"

File: packages/components/src/alignment-matrix-control/index.js
Comment (youknowriad) [inline packages/components/src/alignment-matrix-control/index.js:28]: Do we really need this customized id? Why not just use the regular `useInstanceId`. 
Alternatively, we could support prefixes in `useInstanceId` directly

File: packages/components/src/alignment-matrix-control/index.js
Comment (youknowriad) [inline packages/components/src/alignment-matrix-control/index.js:114]: Very weird to expose both the component and the element? do we really need the element?

File: packages/components/src/utils/hooks.js
Comment (youknowriad) [inline packages/components/src/utils/hooks.js:16]: This doesn't seem used right?

**Review by youknowriad — COMMENTED** [review]
Nice work here. Did you consider whether we want this behavior to be shared across blocks as a support flag or is it something that will stay specific to the Cover block?

File: packages/components/src/alignment-matrix-control/utils.js
Comment (ZebulanStanphill) [inline packages/components/src/alignment-matrix-control/utils.js:29]: This could be simplified a bit:
```suggestion
	for ( const cell of row ) {
		alignments.push( cell );
	}
```

File: packages/components/src/alignment-matrix-control/icon.js
Comment (ZebulanStanphill) [inline packages/components/src/alignment-matrix-control/icon.js:52]: Why not just compare the current `align` to `value`?

**Comment by paaljoachim** [comment]
A cover block can contain multiple inner blocks. For instance a Cover title and a button. How would multiple blocks inside a cover block work together?

File: packages/block-editor/src/components/block-alignment-matrix-toolbar/index.js
Comment (ZebulanStanphill) [inline packages/block-editor/src/components/block-alignment-matrix-toolbar/index.js:31]: I'm pretty sure you shouldn't be using the same `className` for both of these.

File: packages/block-editor/src/components/block-alignment-matrix-toolbar/index.js
Comment (ItsJonQ) [reply packages/block-editor/src/components/block-alignment-matrix-toolbar/index.js:14]: > what level of confidence we have over the naming and the props (API) of this component

I wasn't sure of that distinction either. I've been being conservative, and defaulting to using the `__experimental` prefix.

At being said... My thinking is that I'm currently feeling quite confident. In that case...

* Merge
* Wait a couple of weeks to see if any feedback/bugs pop up
* If nothing, then, safe to remove experimental flag

File: packages/components/src/utils/hooks.js
Comment (ItsJonQ) [reply packages/components/src/utils/hooks.js:16]: Ah! You are correct. It was previously used before we implemented `Composite` from Reakit.
I'll remove it now. Thank you 🙏

File: packages/components/src/alignment-matrix-control/icon.js
Comment (ItsJonQ) [reply packages/components/src/alignment-matrix-control/icon.js:26]: Adding a code comment re: `width` and `height`

File: packages/components/src/alignment-matrix-control/icon.js
Comment (ItsJonQ) [reply packages/components/src/alignment-matrix-control/icon.js:65]: The `value` coming in may not match `align`.
The `getAlignmentIndex` function parses/transforms a `value` prop for accurate reading and comparison purposes.

Originally, the transforming part was more complicated as it allowed for more flexibility in positions, example: both `top center` and `center top` evaluated to the same thing.

Since working with @diegohaz , that logic has been simplified.
It now handles the ability to read `center` as `center center`.

File: packages/components/src/alignment-matrix-control/index.js
Comment (ItsJonQ) [reply packages/components/src/alignment-matrix-control/index.js:28]: > Alternatively, we could support prefixes in useInstanceId directly

I think that may be a good idea! It looks like manually prefixing `instanceId` seems to be the common pattern, example:

```js
	const instanceId = useInstanceId( AnglePickerControl );
	const inputId = `components-angle-picker-control__input-${ instanceId }`;
```

File: packages/components/src/alignment-matrix-control/index.js
Comment (ItsJonQ) [reply packages/components/src/alignment-matrix-control/index.js:124]: I just added this comment:

```
/**
 * Exported rendered icon, which can be used with the <Icon /> component
 * from @wordpress/icons.
 *
 * Example:
 * <Icon icon={ AlignmentMatrixControl.icon } />
 */
```

The idea was to respect how the `<Icon />` component's `icon` prop worked, which expects elements for it to render correctly.

Let me know if this makes sense! I'm open and happy to make changes

Comment (youknowriad) [reply packages/components/src/alignment-matrix-control/index.js:124]: I still find it confusing personally to export the same thing twice. One way to solve this is

```
<Icon icon={ <AlignmentMatrixControl.Icon /> } />
```

Comment (ItsJonQ) [reply packages/components/src/alignment-matrix-control/index.js:124]: We totally could! Does it feel strange that the following are inconsistent? (I'm not opposed, just wanted to double check)

```
<Icon icon={ <AlignmentMatrixControl.Icon /> } />

<Icon icon={ chevronLeft } />
```

Comment (youknowriad) [reply packages/components/src/alignment-matrix-control/index.js:124]: I personally prefer it this way, `<Icon icon={ <AlignmentMatrixControl.Icon /> } />` This is totally reasonable for me. In fact we already do that for somethings. I believe we had something like that for the heading icons at some point (levels)

File: packages/components/src/alignment-matrix-control/icon.js
Comment (youknowriad) [reply packages/components/src/alignment-matrix-control/icon.js:26]: This would solve it right ?

```
<Icon icon={ <AlignmentMatrixControl.Icon size={ something } /> } />
```

File: packages/components/src/alignment-matrix-control/index.js
Comment (ItsJonQ) [reply packages/components/src/alignment-matrix-control/index.js:124]: Cool! Works for me :D. I'll make that update

File: packages/components/src/alignment-matrix-control/icon.js
Comment (ItsJonQ) [reply packages/components/src/alignment-matrix-control/icon.js:26]: Indeed it would! Making those changes now

File: packages/block-editor/src/components/block-alignment-matrix-toolbar/style.scss
Comment (ItsJonQ) [reply packages/block-editor/src/components/block-alignment-matrix-toolbar/style.scss:6]: Currently needed as there's no other way to style the dropdown menu.
It would be lovely to abstract it to Dropdown - perhaps out of scope for this PR?

File: packages/block-editor/src/components/block-alignment-matrix-toolbar/index.js
Comment (ItsJonQ) [reply packages/block-editor/src/components/block-alignment-matrix-toolbar/index.js:31]: It is needed to target the Dropdown (popover) menu, since the menu is being Portal'ed elsewhere in the DOM.

Comment (ZebulanStanphill) [reply packages/block-editor/src/components/block-alignment-matrix-toolbar/index.js:31]: Maybe I'm misunderstanding what you mean, but it should still use a distinct `className`, shouldn't it?

File: packages/components/src/alignment-matrix-control/index.js
Comment (ZebulanStanphill) [inline packages/components/src/alignment-matrix-control/index.js:63]:
```suggestion
		if ( value !== undefined ) {
```

**Comment by ItsJonQ** [comment]
Haii!!

Following up on this one. How are folks feeling? Do you think we're ready to land this?

Thank you!

**Comment by paaljoachim** [comment]
I look forward to testing this out using real world cases.
This should work great with one inner block. 
The tricky part as I see it is when using 2 or more inner blocks and how these would through alignment relate to each other.

**Comment by ItsJonQ** [comment]
> The tricky part as I see it is when using 2 or more inner blocks and how these would through alignment relate to each other.

@paaljoachim Thanks! All inner blocks align together based on the matrix control:

Screen Capture on 2020-05-11 at 10-33-30

Having finer control for how multiple inner items align with each other would be an interesting evolution to this experience!

**Comment by ZebulanStanphill** [comment]
I still have some unaddressed comments on the code, but aside from those, this seems good to go.

File: packages/block-library/src/cover/editor.scss
Comment (ItsJonQ) [reply packages/block-library/src/cover/editor.scss]: Thank you for catching that. I'll move this one 👍

File: packages/block-editor/src/components/block-alignment-matrix-toolbar/index.js
Comment (ItsJonQ) [reply packages/block-editor/src/components/block-alignment-matrix-toolbar/index.js:31]: I suppose it could. I'll make that adjustment

File: packages/block-library/src/cover/style.scss
Comment (ItsJonQ) [reply packages/block-library/src/cover/style.scss:161]: Indeed it does! I can't remember why I had kept it outside that scope. In any case, I'll make that update.

Thanks!

**Comment by ItsJonQ** [comment]
@ZebulanStanphill Thanks for following up! I went over your items and pushed the updates. Let me know if there's anything I may have missed! 🙏

**Comment by ZebulanStanphill** [comment]
The dropdown for the matrix control lacks G2 borders, but I think that might just be because the branch needs a rebase? It might also not be a problem with this PR specifically, but a general styling issue that needs to be fixed.

image

File: packages/components/src/alignment-matrix-control/utils.js
Comment (mtias) [inline packages/components/src/alignment-matrix-control/utils.js:18]: nit: but I think we should `middle` for the vertical axis and `center` for the horizontal to avoid this "center center" repetition.

**Review by mtias — CHANGES_REQUESTED** [review]

File: packages/components/src/alignment-matrix-control/utils.js
Comment (ZebulanStanphill) [reply packages/components/src/alignment-matrix-control/utils.js:18]: As far as I'm aware, there's no difference in meaning between "middle" and "center", or at least not relating to axis. We could just have the displayed title be simply "Center".

Comment (ZebulanStanphill) [inline packages/components/src/alignment-matrix-control/utils.js:18]:
```suggestion
	'center center': __( 'Center' ),
```

File: packages/block-editor/src/components/block-alignment-matrix-toolbar/index.js
Comment (mtias) [reply packages/block-editor/src/components/block-alignment-matrix-toolbar/index.js:14]: Agreed, I think we should possibly apply this to the _Group_ block once merged and see if it works fine before considering removing the _experimental prefix.

**Review by mtias — APPROVED** [review]

**Comment by ItsJonQ** [comment]
@mtias Thank you for reviewing! I've rebased with latest `master`. Will merge this up once Travis is green and happy :)

**Comment by ItsJonQ** [comment]
Yay! Merging! 😊

**Comment by mtias** [comment]
Nice work, really like this one :)

**Comment by paaljoachim** [comment]
Hey

I added a background image. A title and a button below it. But I am having a hard time aligning the title and the button to the middle of the Cover Block.

Title and button have no alignment selected.

Screen Shot 2020-06-08 at 22 41 17


Aligning the title and the button to center looks like this:
(Grid is center aligned.) Here I have selected both the button and the title.

Screen Shot 2020-06-08 at 22 45 00

I am guessing a new issue is needed to address the centering of the Button.

**Comment by rmorse** [comment]
Found a little bug when opening the control, focus is triggered (and so is onChange) - the fix is here  https://github.com/WordPress/gutenberg/pull/27945

**Pull request merged**
